### PR TITLE
rocm-related: 5.3.3 → 5.4.0

### LINKED ISSUE #203949: rocm-opencl-runtime: "Output limit exceeded" on Hydra
https://hydra.nixos.org/build/200341028

### Notify maintainers
@acowley @lovesegfault @Flakebi

### PR DESCRIPTION
**Pull request by Madouura**
###### Description of changes
Tracking: #197885
Manually updated ``rocthrust`` and ``tensile`` because the release order is messed up...
Also fixed an issue with the update scripts that could have resulted in a bunch of the rocm packages being marked as broken due to the rocm version check on most of them. (5.4 != 5.4.0)
~~Will update the section below once I'm finished building.~~
Fixes #203949

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by Madouura** [comment]
I probably should have made this a draft.
Don't merge yet.

**Comment by Madouura** [comment]
Result of `nixpkgs-review pr 204378` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>32 packages built:</summary>
  <ul>
    <li>blender-hip</li>
    <li>clang-ocl</li>
    <li>composable_kernel</li>
    <li>hip</li>
    <li>hipcub</li>
    <li>hipsparse</li>
    <li>llvmPackages_rocm.clang</li>
    <li>llvmPackages_rocm.clangNoCompilerRt</li>
    <li>llvmPackages_rocm.llvm</li>
    <li>miopen (miopen-hip)</li>
    <li>miopen-opencl</li>
    <li>miopengemm</li>
    <li>rccl</li>
    <li>rocblas</li>
    <li>rocclr</li>
    <li>rocfft</li>
    <li>rocm-cmake</li>
    <li>rocm-comgr</li>
    <li>rocm-device-libs</li>
    <li>rocm-opencl-icd</li>
    <li>rocm-opencl-runtime</li>
    <li>rocm-runtime</li>
    <li>rocm-smi</li>
    <li>rocm-thunk</li>
    <li>rocminfo</li>
    <li>rocmlir</li>
    <li>rocprim</li>
    <li>rocrand</li>
    <li>rocsparse</li>
    <li>rocthrust</li>
    <li>rocwmma</li>
    <li>tensile</li>
  </ul>
</details>

**Comment by Madouura** [comment]
Ready.

File: pkgs/development/compilers/hip/default.nix [hunk 1]
```diff
@@ -172,6 +169,14 @@ stdenv.mkDerivation (finalAttrs: {
     #!/usr/bin/env nix-shell
     #!nix-shell -i bash -p curl jq common-updater-scripts nix-prefetch-github
     version="$(curl ''${GITHUB_TOKEN:+"-u \":$GITHUB_TOKEN\""} -sL "https://api.github.com/repos/ROCm-Developer-Tools/HIP/tags" | jq '.[].name | split("-") | .[1] | select( . != null )' --raw-output | sort -n | tail -1)"
+
+    IFS='.'
+    read -a version_arr <<< "$version"
+
+    if [ "''${#version_arr[*]}" == 2 ]; then
+      version="''${version}.0"
+    fi
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/hip/default.nix:178]: We probably want to restore IFS for the rest of the build

**Review by SuperSandro2000 — COMMENTED** [review]
We should do a generic updater for rocm instead of copying the changes to all packages.

File: pkgs/development/compilers/hip/default.nix [hunk 2]
```diff
@@ -172,6 +169,14 @@ stdenv.mkDerivation (finalAttrs: {
     #!/usr/bin/env nix-shell
     #!nix-shell -i bash -p curl jq common-updater-scripts nix-prefetch-github
     version="$(curl ''${GITHUB_TOKEN:+"-u \":$GITHUB_TOKEN\""} -sL "https://api.github.com/repos/ROCm-Developer-Tools/HIP/tags" | jq '.[].name | split("-") | .[1] | select( . != null )' --raw-output | sort -n | tail -1)"
+
+    IFS='.'
+    read -a version_arr <<< "$version"
```
Comment (Atemu) [inline pkgs/development/compilers/hip/default.nix:174]: My bash is a bit rusty but doesn't

```suggestion
    IFS='.' read -a version_arr <<< "$version"
```

work?

**Comment by Madouura** [comment]
> We should do a generic updater for rocm instead of copying the changes to all packages.

Alright, I'll go ahead and incorporate the changes from #203949 into that since it'll be easier for all of us then.

File: pkgs/development/compilers/hip/default.nix (hunk 2 above)
Comment (Madouura) [reply pkgs/development/compilers/hip/default.nix:174]: Mine is too, AFAIK we won't have to ``unset IFS`` if we do this.

File: pkgs/development/compilers/hip/default.nix (hunk 1 above)
Comment (Madouura) [reply pkgs/development/compilers/hip/default.nix]: AFAIK putting IFS before the command should work, but my bash is rusty so let me know if that isn't working.

**Comment by Madouura** [comment]
Haven't rebuilt everything, but from what I can tell there are no glaring ``nix-build`` errors for evaluation.
I have ran the update scripts a few times with various changes to make sure that's working, and it is.

**Comment by Madouura** [comment]
Result of `nixpkgs-review pr 204378` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>31 packages built:</summary>
  <ul>
    <li>blender-hip</li>
    <li>clang-ocl</li>
    <li>composable_kernel</li>
    <li>hip</li>
    <li>hipcub</li>
    <li>hipsparse</li>
    <li>llvmPackages_rocm.clang</li>
    <li>llvmPackages_rocm.clangNoCompilerRt</li>
    <li>llvmPackages_rocm.llvm</li>
    <li>miopen (miopen-hip)</li>
    <li>miopen-opencl</li>
    <li>miopengemm</li>
    <li>rccl</li>
    <li>rocblas</li>
    <li>rocclr</li>
    <li>rocm-cmake</li>
    <li>rocm-comgr</li>
    <li>rocm-device-libs</li>
    <li>rocm-opencl-icd</li>
    <li>rocm-opencl-runtime</li>
    <li>rocm-runtime</li>
    <li>rocm-smi</li>
    <li>rocm-thunk</li>
    <li>rocminfo</li>
    <li>rocmlir</li>
    <li>rocprim</li>
    <li>rocrand</li>
    <li>rocsparse</li>
    <li>rocthrust</li>
    <li>rocwmma</li>
    <li>tensile</li>
  </ul>
</details>

**Comment by Madouura** [comment]
cc @SuperSandro2000
Are there any further issues I need to address before we get this merged?

File: pkgs/development/libraries/hipcub/default.nix [hunk 3]
```diff
@@ -16,9 +16,12 @@
 }:
 
 # CUB can also be used as a backend instead of rocPRIM.
-stdenv.mkDerivation (finalAttrs: {
+let
+  srcOwner = "ROCmSoftwarePlatform";
+  srcRepo = "hipCUB";
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/hipcub/default.nix:21]: We don't need to do this. You should be able to access owner and repo like ``src.repo`` and ``src.owner``.

Comment (Madouura) [reply pkgs/development/libraries/hipcub/default.nix:21]: I tried multiple times, it's not possible.
Both with ``finalAttrs`` and trying with the normal ``rec``.

Comment (Madouura) [reply pkgs/development/libraries/hipcub/default.nix:21]: See this line for why: https://github.com/NixOS/nixpkgs/blob/32df8276ec827f7a3e980f190eb82051eb0ca68f/pkgs/build-support/fetchgithub/default.nix#L25

Comment (Madouura) [reply pkgs/development/libraries/hipcub/default.nix:21]: This is the exact error, sans variations for ``repo``:
```console
error: attribute 'owner' missing

       at /home/mado/Documents/Development/nixpkgs/pkgs/development/libraries/clang-ocl/default.nix:39:13:

           38|     name = finalAttrs.pname;
           39|     owner = finalAttrs.src.owner;
             |             ^
           40|     repo = finalAttrs.src.repo;
(use '--show-trace' to show detailed location information)
```

Comment (Madouura) [reply pkgs/development/libraries/hipcub/default.nix:21]: One option we have is something like this:
```nix
stdenv.mkDerivation (finalAttrs: {
  pname = "clang-ocl";
  version = "5.4.0";

  src = fetchFromGitHub {
    owner = finalAttrs.passthru.owner;
    repo = finalAttrs.passthru.repo;
    rev = "rocm-${finalAttrs.version}";
    hash = "sha256-uMSvcVJj+me2E+7FsXZ4l4hTcK6uKEegXpkHGcuist0=";
  };

  passthru = {
    owner = "RadeonOpenCompute";
    repo = "clang-ocl";
    
    updateScript = rocmUpdateScript {
      inherit (finalAttrs) pname;
      inherit (finalAttrs.passthru) owner repo;
    };
  };
})
```
...And changing ``name`` in the update script to ``pname`` so we can just easily inherit it, but honestly it seems pointless unless we absolutely need to retain information that we somehow can't with ``let ... in``

Comment (SuperSandro2000) [reply pkgs/development/libraries/hipcub/default.nix:21]: I think we should rather drop owner and repo from this list https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/fetchgithub/default.nix#L25

**Comment by Madouura** [comment]
Looks like the ``rocm-llvm`` was a bit more of a rabbit hole than I thought.
I'm saving that for a later PR in maybe a week or so.
This should fix the immediate issues.

**Comment by Madouura** [comment]
Almost forgot to revert the change that's been causing hydra errors.

**Comment by lovesegfault** [comment]
Result of `nixpkgs-review pr 204378` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>31 packages built:</summary>
  <ul>
    <li>blender-hip</li>
    <li>clang-ocl</li>
    <li>composable_kernel</li>
    <li>hip</li>
    <li>hipcub</li>
    <li>hipsparse</li>
    <li>llvmPackages_rocm.clang</li>
    <li>llvmPackages_rocm.clangNoCompilerRt</li>
    <li>llvmPackages_rocm.llvm</li>
    <li>miopen (miopen-hip)</li>
    <li>miopen-opencl</li>
    <li>miopengemm</li>
    <li>rccl</li>
    <li>rocblas</li>
    <li>rocclr</li>
    <li>rocm-cmake</li>
    <li>rocm-comgr</li>
    <li>rocm-device-libs</li>
    <li>rocm-opencl-icd</li>
    <li>rocm-opencl-runtime</li>
    <li>rocm-runtime</li>
    <li>rocm-smi</li>
    <li>rocm-thunk</li>
    <li>rocminfo</li>
    <li>rocmlir</li>
    <li>rocprim</li>
    <li>rocrand</li>
    <li>rocsparse</li>
    <li>rocthrust</li>
    <li>rocwmma</li>
    <li>tensile</li>
  </ul>
</details>

**Pull request merged**